hash-table/chaining.js

```diff
@@ -0,0 +1,147 @@
+const { Node, DoublyLinkedList } = require('../linked-list/doubly-linked-list')
+
+const DEFAULT_CAPACITY = 8
+class ChainingHashTable {
+  constructor() {
+    this.arr = new Array(DEFAULT_CAPACITY)
+    this.capacity = DEFAULT_CAPACITY
+    this.size = 0
+  }
+
+  insert(value, key) {
+    const index = this.hash(key)
+    if (!this.arr[index]) {
+      const linkList = new DoublyLinkedList()
+      linkList.pushFront(value, key)
+      this.arr[index] = linkList
+      this.size++
+      if (this.size === this.capacity) {
+        this.grow()
+      }
+      return
+    }
+
+    const isExisting = this.search(key)
+    if (!isExisting) {
+      this.arr[index].pushBack(value, key)
+      return
+    }
+    let cur = this.arr[index].tail
+    while (cur) {
+      if (cur.key === key) {
+        cur.value = value
+        break
+      }
+      cur = cur.prev
+    }
+  }
+
+  delete(key) {
+    const index = this.hash(key)
+    
+    if (!this.arr[index]) return
+    if (this.arr[index].length === 1) {
+      this.arr[index] = undefined
+      this.size--
+      if (this.capacity > DEFAULT_CAPACITY && this.size <= Math.floor(this.capacity / 4)) {
```

Comment:
There's still one more magic number here, which is 4. Ideally, you should check if the *load factor* of the Hash Table is less than a number, if yes then shrink. Load factor varies from 0 to 1. You can read more about it [here](https://en.wikipedia.org/wiki/Hash_table#Key_statistics).